### PR TITLE
Fixed #1044

### LINKED ISSUE #1044: AADL Package Wizard Allows Creating Package with Conflicting Names
The AADL package wizard does not check if there is a package with the same name referencable by the project. If there is a conflicting package in a separate folder, the wizard will complete, but the new package will have an error.

### PR DESCRIPTION
**Pull request by AaronGreenhouse**
Fixed this.

### CODE REVIEW

File: org.osate.ui/src/org/osate/ui/wizards/NewAadlPackageWizard.xtend [hunk 1]
```diff
@@ -67,6 +78,26 @@ class NewAadlPackageWizard extends AbstractNewFileWizard {
...
+		/* Parent might be a Project, which causes problems below, so let's append
+		 * a bogus folder to it.
+		 */
+		val IFolder fakeFolder = parent.getFolder(Path.forPosix(".fake"))
+		val Resource rsrc = OsateResourceUtil.getResource(fakeFolder)
+		val scope = globalScopeProvider.getScope(
+			rsrc, Aadl2Package.eINSTANCE.getPackageRename_RenamedPackage(), null
+		)
+		val segments = packageName.split("::")
+		val qualifiedName = QualifiedName.create(segments);
```
Comment (lwrage) [inline org.osate.ui/src/org/osate/ui/wizards/NewAadlPackageWizard.xtend]: Better inject an IQualifiedNameConverter and use its toQualifiedName method.

**Comment by AaronGreenhouse** [comment]
Fixed to use an injected IQualifedNameConverter

**Comment by lwrage** [comment]
Could you delete the commented out lines in NewAadlPackageWizard.xtend, please?

**Comment by AaronGreenhouse** [comment]
done

File: org.osate.ui/src/org/osate/ui/wizards/NewAadlPackageWizard.xtend [hunk 2]
```diff
@@ -67,6 +80,25 @@ class NewAadlPackageWizard extends AbstractNewFileWizard {
...
+		 * a bogus folder to it.
+		 */
+		val IFolder fakeFolder = parent.getFolder(Path.forPosix(".fake"))
+		val Resource rsrc = OsateResourceUtil.getResource(fakeFolder)
+		val scope = globalScopeProvider.getScope(
+			rsrc, Aadl2Package.eINSTANCE.getPackageRename_RenamedPackage(), null
+		)
+		val qualifiedName = qNameConverter.toQualifiedName(packageName);
+		if (scope.getSingleElement(qualifiedName) !== null) {
+			return "Package '" + packageName + "' already exists in scope."
```
Comment (lwrage) [inline org.osate.ui/src/org/osate/ui/wizards/NewAadlPackageWizard.xtend]: The error message is correct but not very useful. It should read: "Package xyz already exists in project abc (file fgh.aadl)." This helps the user find the other package.

**Comment by lwrage** [comment]
This leaves the situation described in https://github.com/osate/osate2-core/issues/1044#issuecomment-370920533 unhandled, doesn't it? It may be a good idea to address that as well.

**Comment by AaronGreenhouse** [comment]
So what do we want to do if 
- `project2` contains package `p2`
- `project2` depends on `project1`
- And the user wants to create a package `p2` in `project1`?

This doesn't create any problems for `project1`.  I'm not sure it should, but yes, it would be a courtesy to alert that it creates problems in `project2`.

**Comment by lwrage** [comment]
Obviously that requires checking all projects that depend on project1.

File: org.osate.ui/src/org/osate/ui/wizards/NewAadlPackageWizard.xtend [hunk 3]
```diff
@@ -68,6 +81,50 @@ class NewAadlPackageWizard extends AbstractNewFileWizard {
...
+			 */
+			for (user : parent.getProject().referencingProjects) {
+				val found2 = findPackageInScope(user, packageName)
+				if (found2 !== null) {
+					val foundFile = OsateResourceUtil.getOsateIFile(found2.EObjectURI)
+					val foundProject = foundFile.getProject()
+					if (foundProject != user) {
+						return "Package '" + packageName + "' already exists in project '" + foundProject.name + "' that is depended on by project '" + user.name + "' that depends on the selected project: '" + foundFile.projectRelativePath + "'"
+					} else {
+						return "Package '" + packageName + "' already exists in project '" + user.name + "' that depends on the selected project: '" + foundFile.projectRelativePath + "'"
```
Comment (sprocter) [inline org.osate.ui/src/org/osate/ui/wizards/NewAadlPackageWizard.xtend]: Example comment.

File: org.osate.ui/src/org/osate/ui/wizards/AbstractNewFileWizard.xtend [hunk 4]
```diff
@@ -229,7 +230,15 @@ abstract class AbstractNewFileWizard extends Wizard implements INewWizard {
 			wizardPage.pageComplete = false
 		])
 	}
-			
+	
+	def protected String validateFileName(IContainer parent, String name) {
+		if (file.exists) {
```
Comment (AaronGreenhouse) [inline org.osate.ui/src/org/osate/ui/wizards/AbstractNewFileWizard.xtend]: Added a new method validateFileName() that is invoked by validate().  This method is meant to be overridden by subclasses.  The initial implementation mimics was in validate().

File: org.osate.ui/src/org/osate/ui/wizards/AbstractNewFileWizard.xtend [hunk 5]
```diff
@@ -217,8 +217,9 @@ abstract class AbstractNewFileWizard extends Wizard implements INewWizard {
 							fieldLabel + " cannot be empty."
 						} else if (!fieldValidators.get(fieldLabel).apply(field.text)) {
 							'''The «fieldLabel» '«field.text»' is not valid.'''
-						} else if (fieldLabel == fields.keySet.head && file.exists) {
-							"'" + getFileName(fields.values.head.text) + "' already exists."
+						} else if (fieldLabel == fields.keySet.head) {
```
Comment (AaronGreenhouse) [inline org.osate.ui/src/org/osate/ui/wizards/AbstractNewFileWizard.xtend]: Here is where the call to new validateFileName() method is inserted.  It replaces the `file.exists` check that was previously in the conditional.

File: org.osate.ui/src/org/osate/ui/wizards/NewAadlPackageWizard.xtend [hunk 6]
```diff
@@ -68,6 +81,50 @@ class NewAadlPackageWizard extends AbstractNewFileWizard {
 		]
 	}
 	
+	def private findPackageInScope(IContainer parent, String packageName) {
```
Comment (AaronGreenhouse) [inline org.osate.ui/src/org/osate/ui/wizards/NewAadlPackageWizard.xtend]: New method for looking up a package name in the scope of the project that *is* or contains `parent`.

File: org.osate.ui/src/org/osate/ui/wizards/NewAadlPackageWizard.xtend [hunk 7]
```diff
@@ -68,6 +81,50 @@ class NewAadlPackageWizard extends AbstractNewFileWizard {
...
+			/* See if the package exists in scope in any project that depends on the current project.
+			 * This isn't an error for the current project, but it will mess up other projects.
+			 */
+			for (user : parent.getProject().referencingProjects) {
+				val found2 = findPackageInScope(user, packageName)
+				if (found2 !== null) {
+					val foundFile = OsateResourceUtil.getOsateIFile(found2.EObjectURI)
+					val foundProject = foundFile.getProject()
+					if (foundProject != user) {
+						return "Package '" + packageName + "' already exists in project '" + foundProject.name + "' that is depended on by project '" + user.name + "' that depends on the selected project: '" + foundFile.projectRelativePath + "'"
```
Comment (AaronGreenhouse) [inline org.osate.ui/src/org/osate/ui/wizards/NewAadlPackageWizard.xtend]: The most round-about case.  The named package exists in project `C` that project `B` depends on, and proejct `B` also depends on the currently selected project.  Not a problem for the current project, but breaks project `B`.

File: org.osate.ui/src/org/osate/ui/wizards/AbstractNewFileWizard.xtend [hunk 8]
```diff
@@ -229,7 +230,15 @@ abstract class AbstractNewFileWizard extends Wizard implements INewWizard {
 			wizardPage.pageComplete = false
 		])
 	}
-			
+	
+	def protected String validateFileName(IContainer parent, String name) {
+		if (file.exists) {
+			return "'" + getFileName(name) + "' already exists.";
```
Comment (lwrage) [inline org.osate.ui/src/org/osate/ui/wizards/AbstractNewFileWizard.xtend]: return unnecessary

File: org.osate.ui/src/org/osate/ui/wizards/AbstractNewFileWizard.xtend [hunk 9]
```diff
@@ -217,8 +217,9 @@ abstract class AbstractNewFileWizard extends Wizard implements INewWizard {
 							fieldLabel + " cannot be empty."
 						} else if (!fieldValidators.get(fieldLabel).apply(field.text)) {
 							'''The «fieldLabel» '«field.text»' is not valid.'''
-						} else if (fieldLabel == fields.keySet.head && file.exists) {
-							"'" + getFileName(fields.values.head.text) + "' already exists."
+						} else if (fieldLabel == fields.keySet.head) {
+							val parentContainer = folderViewer.structuredSelection.firstElement as IContainer
+							return validateFileName(parentContainer, fields.values.head.text)
```
Comment (lwrage) [inline org.osate.ui/src/org/osate/ui/wizards/AbstractNewFileWizard.xtend]: unnecessary "return"

File: org.osate.ui/src/org/osate/ui/wizards/NewAadlPackageWizard.xtend [hunk 10]
```diff
@@ -68,6 +81,50 @@ class NewAadlPackageWizard extends AbstractNewFileWizard {
 		]
 	}
 	
+	def private findPackageInScope(IContainer parent, String packageName) {
+		/* Parent might be a Project, which causes problems below, so let's append
+		 * a bogus folder to it.
+		 */
+		val IFolder fakeFolder = parent.getFolder(Path.forPosix(".fake"))
+		val Resource rsrc = OsateResourceUtil.getResource(fakeFolder)
+		val scope = globalScopeProvider.getScope(rsrc, Aadl2Package.eINSTANCE.getPackageRename_RenamedPackage(), null)
```
Comment (lwrage) [inline org.osate.ui/src/org/osate/ui/wizards/NewAadlPackageWizard.xtend]: Use reference from with clause, instead, to find property sets and packages.

**Comment by AaronGreenhouse** [comment]
Per yesterday's code review, fixed the "return" statements and added a large comment about what error cases are being caught.

**Comment by AaronGreenhouse** [comment]
Fixed to check for both property sets and packages

**Pull request merged**